drivers/misc/morse_code/morse_code.c

```diff
@@ -0,0 +1,406 @@
+#include <zephyr/drivers/misc/morse_code/morse_code.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(morse_code, CONFIG_MORSE_CODE_LOG_LEVEL);
+
+#define DT_MORSE_PROP(inst, prop) DT_INST_PROP(inst, prop)
+
+enum morse_code_stream_fsm {
+	MORSE_CODE_STREAM_FSM_LOAD,
+	MORSE_CODE_STREAM_FSM_TX,
+};
+
+enum morse_code_bit_fsm {
+	MORSE_CODE_BIT_FSM_LOAD_SYMBOL,
+	MORSE_CODE_BIT_FSM_WAIT_BLANK_PHASE,
+};
+
+enum morse_code_prosigns {
+	MORSE_CODE_PROSIGN_LETTER_SPACE = 0x40,
+	MORSE_CODE_PROSIGN_CT_START_TX,
+	MORSE_CODE_PROSIGN_AR_END_TX,
+};
+
+struct morse_code_callback_data {
+	morse_callback_handler_t *callback;
+	void *ctx;
+	int status;
+};
+
+struct morse_code_data {
+	struct k_work cb_work;
+	struct morse_code_callback_data cb_info;
+	struct counter_top_cfg dot_tick;
+	const uint8_t *data;
+	size_t data_idx;
+	size_t data_size;
+	enum morse_code_stream_fsm code_fsm;
+	enum morse_code_bit_fsm bit_fsm;
+	int code_idx;
+	int code_bit;
+};
+
+struct morse_code_config {
+	const struct device *const timer;
+	const struct gpio_dt_spec led;
+	uint32_t speed;
+};
+
+/*
+ * The bit encoding follows the ITU-R M.1677-1 for spacing and length of the
+ * signals were:
+ *
+ * 1: The first MSB nimble store the length in bits of the symbol. This is used
+ * to optimize the "search first 1 bit" in the symbol bit stream.
+ *
+ * 2:
+ *	The '.' (dot) is encoded as 1 bit with value 1
+ * 3.1:
+ *	A dash is equal to three dots. The 3 dots are 3 consecutive bits with
+ *	value 1.
+ * 3.2:
+ *	The space between the signals forming the same letter is equal to one
+ *	dot. The encoding is made using 1 bit with value 0.
+ * 3.3:
+ *	The space between two letters is equal to three dots. This is equivalent
+ *	to 3 consecutive bits with value 0.
+ * 3.4:
+ *	The space between two words is equal to seven dots. This is equivalent
+ *	to 7 consecutive bits with value 0.
+ *
+ * Note: The space between letters and words are implemented in the function
+ * and are not encoded in the morse code table.
+ *
+ */
+static const uint32_t morse_code_symbols[] = {
+/*      <encoding>,    <meaning> <morse code> <bit encoding>
+ *
+ * Reserved
+ *	0x0f0075d7,    /CT        -.-.-       111010111010111                  111 0101 1101 0111
+ *	0x0d00175d,    /AR        .-.-.       1011101011101                      1 0111 0101 1101
+ *	0x0f0055d7,    /VA        ...-.-      101010111010111                  101 0101 1101 0111
+ *	0x090001d7,     K         -.-         111010111                               1 1101 0111
+ *	0x0f0075dd,     KN        -.--.       111010111011101                  111 0101 1101 1101
+ *	0x0b0005d5,    wait       .-...       10111010101                           101 1101 0101
+ *	0x0b00055d,    understood ...-.       10101011101                           101 0101 1101
+ *	0x0f005555,    error      ........    101010101010101                  101 0101 0101 0101
+ */
+
+/*
+ *      <encoding>,    <symbol> <morse code> <bit encoding>
```

Comment:
More detail about this "_encoding_" would be nice...

From reading the rest of the code, it seems you've split the `uint32_t` into two - `[31:24]` is the length, and `[23:0]` is the symbol's morse encoding, correct?

Reply:
I added a few more info. Please, check and give feedback.

Reply:
"_start bit_" vs "_index of first bit to transmit_" (or something more related to length)

I'm also wondering if a macro would be nice, letting us use decimal start index, and hex / binary bit sequence (and somewhat "_self-documenting_" too), e.g:

```c
#define MORSE_SYM(start_idx, bit_stream) (0 \
    | ((start_idx << 24) & 0xff000000) \
    | ((bit_stream << 0) & 0x00ffffff))

static const uint32_t morse_code_symbols[] = {
    MORSE_SYM( 7,               0b0000000), /* word space [3.4]  */
    MORSE_SYM(19,   0b1110101110101110111), /*     !     -.-.--  */
    MORSE_SYM(15,       0b101110101011101), /*     "     .-..-.  */
    /* ... */
    MORSE_SYM(13,         0b1011101011101), /* /AR        .-.-.  */
};
```

Reply:
fyi https://github.com/zephyrproject-rtos/zephyr/pull/61800 (but this could be a justifiable exception I guess)

Reply:
@kartben thanks for linking the PR! ... I hadn't realized they weren't part of the standard, but do remember them not working on a toolchain long (long) ago!